api/v1alpha1/s3bucket_types.go

```diff
@@ -0,0 +1,132 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// S3BucketParameters define the desired state of a Cloudscale S3 Bucket
+// https://www.cloudscale.ch/en/api/v1#objects-users
+type S3BucketParameters struct {
+	// NameFormat specifies the name of the external S3Bucket instance. The
+	// first instance of the string '%s' will be replaced with the Kubernetes
+	// UID of this S3Bucket. Omit this field to use the UID alone as the name.
+	// +optional
+	NameFormat string `json:"nameFormat,omitempty"`
+
+	// Tags are optional key, value pairs to add to an S3 bucket
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
+}
+
+// S3BucketSpec defines the desired state of S3Bucket
+type S3BucketSpec struct {
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+	S3BucketParameters           `json:",inline"`
+}
+
+// S3BucketStatus defines the observed state of S3Bucket
+type S3BucketStatus struct {
+	runtimev1alpha1.ResourceStatus `json:",inline"`
+
+	Status       string `json:"status,omitempty"`
+	ObjectUserID string `json:"objectUserId,omitempty"`
+	BucketName   string `json:"bucketName,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// S3Bucket is the Schema for the s3buckets API
+// +kubebuilder:printcolumn:name="CLASS",type="string",JSONPath=".spec.classRef.name"
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+type S3Bucket struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   S3BucketSpec   `json:"spec,omitempty"`
+	Status S3BucketStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// S3BucketList contains a list of S3Bucket
+type S3BucketList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []S3Bucket `json:"items"`
+}
+
+// An S3BucketClassSpecTemplate is a template for the spec of a dynamically
+// provisioned S3Bucket.
+type S3BucketClassSpecTemplate struct {
+	runtimev1alpha1.ClassSpecTemplate `json:",inline"`
+	S3BucketParameters                `json:",inline"`
```

Comment:
Another "newer" pattern we have begun to standardize around is call the embedded parameters struct `forProvider`. Once again not required by any means, but it may be beneficial in standardizing around other stacks patterns. Then, similarly, we have an `S3BucketObservation` struct that we embed in the `status` as `atProvider`.

Example Ref: https://github.com/crossplaneio/stack-aws/blob/81f612019be228db4c093b6879f4485d61ef9d08/apis/database/v1beta1/rdsinstance_types.go#L818

Docs Ref: https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#owner-based-struct-tree